api/controllers/StatsController.js

```diff
@@ -6,6 +6,40 @@ dayjs.extend(isBetween);
 /////////////
 // Helpers //
 /////////////
+/**
+ * Find the season in which a given timestamp occurred using binary search
+ * @param {*} sortedSeasons - Seasons sorted (OLDEST FIRST)
+ * @param {*} timeStamp - Desired time in millis since epoch
+ * @returns the season | null if none is found
+ */
+function getSeasonFromTimeStamp(sortedSeasons, timeStamp) {
+  let currentIndex;
+  let minIndex = 0;
+  let maxIndex = sortedSeasons.length - 1;
+  // Sanity check in case list is unsorted or start & endTimes are out of whack
+  let numTries = 0;
+  while (numTries < sortedSeasons.length) {
+    currentIndex = Math.floor((minIndex + maxIndex) / 2);
+    const currentSeason = sortedSeasons[currentIndex];
+    if (timeStamp >= currentSeason.startTime && timeStamp <= currentSeason.endTime) {
+      return currentSeason;
+    }
+    // If min = max and we haven't found it, there is no matching season
+    if (minIndex === maxIndex) {
+      return null;
+    }
+    // Target time is before current season
+    if (timeStamp < currentSeason.startTime) {
```

Comment:
Would it be better to allow a query with timestamp on the records so we don't need to do that logic in JavaScript?

Reply:
Actually there is overhead that comes with each individual query e.g. network latency since the database and app server live on different vm's and it is generally more performant to do fewer, batched queries for more data then process it with serverside code than it is to repeatedly make smaller queries that each require minimal serverside processing.

As a rule of thumb, rather than querying from inside a loop, you want to query once and loop through the results. That's why this controller method queries for the seasons once at the top of the api func and then passes the array of seasons as an argument here to prevent repeated queries